### PR TITLE
[RFC] Display implicit args in header

### PR DESCRIPTION
**Pull request by blueyed**
TODO:

- [ ] changelog

### CODE REVIEW

File: testing/test_config.py [hunk 1]
```diff
@@ -1176,3 +1176,36 @@ def pytest_addoption(parser):
...
+    monkeypatch.setenv("PYTEST_ADDOPTS", "-v")
+    result = testdir.runpytest(str(p))
+    result.stdout.fnmatch_lines(
+        ["implicit args: \"-k 'foo and bar'\" (addopts config), '-v' (PYTEST_ADDOPTS)"]
+    )
+
+    ini.remove()
+    monkeypatch.setenv("PYTEST_ADDOPTS", "-v -k 'bar'")
+    result = testdir.runpytest(str(p))
+    result.stdout.fnmatch_lines(["implicit args: '-v -k bar' (PYTEST_ADDOPTS)"])
```
Comment (blueyed) [inline testing/test_config.py]: Maybe unit-tests would be better/faster here, except for one end-to-end, but waiting for general feedback first.

**Comment by blueyed** [comment]
This is useful for when looking at CI logs and new projects in general.
I've also thought about displaying the effective args altogehter (for copy'n'paste), but that might be rather logs given file args etc.
It would also be nice if the implicit args could be displayed so that the are copy'n'pastable in a single go, but then you cannot display where they are coming from, except for when using bold attributes for one of them maybe.

Maybe it should also get displayed with the first line ("rootdir: …") at the end, to save a line with wide displays.

**Comment by RonnyPfannschmidt** [comment]
This one will need a fat warning as it may expose credentials in bad setups

**Comment by blueyed** [comment]
Good point!
Maybe it should only show source (names) by default ("addopts config", "PYTEST_ADDOPTS"), but then everything with `--traceconfig`?
OTOH this would not help when looking at CI logs for failed jobs.
But certainly an option to opt out would be good - not sure if it makes sense to bind it to `verbose`?!

**Comment by blueyed** [comment]
Are you OK with displaying additional arg sources by name as a first step?
And is it OK to display it with `--traceconfig` (have not really looked into how this works)?

File: src/_pytest/terminal.py [hunk 2]
```diff
@@ -597,6 +598,16 @@ def pytest_report_header(self, config):
         plugininfo = config.pluginmanager.list_plugin_distinfo()
         if plugininfo:
             result.append("plugins: %s" % ", ".join(_plugin_nameversions(plugininfo)))
+
+        if config._implicit_args:
+            result.append(
+                "implicit args: %s"
+                % ", ".join(
+                    "%r (%s)" % (" ".join([shell_quote(x) for x in values]), desc)
```
Comment (nicoddemus) [inline src/_pytest/terminal.py]: I suggest to break this expression in some variables, as it is formatted it is quite hard to parse.

Comment (nicoddemus) [reply src/_pytest/terminal.py]: Perhaps we can have one section per line? For example at work we set `PYTEST_ADDOPTS=--max-slave-restart=10 --timeout=60 -ra --color=yes --replay-record-dir=build/tests/replay`, which can get quite big.

File: src/_pytest/terminal.py [hunk 3]
```diff
@@ -597,6 +598,16 @@ def pytest_report_header(self, config):
         plugininfo = config.pluginmanager.list_plugin_distinfo()
         if plugininfo:
             result.append("plugins: %s" % ", ".join(_plugin_nameversions(plugininfo)))
+
+        if config._implicit_args:
+            result.append(
```
Comment (nicoddemus) [inline src/_pytest/terminal.py]: I suggest we also introduce a new option, `show_implicit_arguments` defaulting to True, so users can turn this off in case they use secrets in their options or their output is too cluttered.

**Comment by nicoddemus** [comment]
Hey @blueyed,

I'm closing this for now to clean up our PR queue a bit, feel free to re-open this if you want to tackle this again.
